File: bindings/elixir/lib/causalontology/jcs.ex
defmodule Causalontology.Jcs do
  @moduledoc """
  RFC 8785 (JSON Canonicalization Scheme) serialization.

  Mirrors `bindings/python/causalontology/canonical.py`'s `_jcs` exactly:
  sorted keys (code-point order, which equals UTF-16 code-unit order for the
  ASCII keys the standard uses), minimal string escaping with lowercase
  `\\u00xx` for control characters, and ECMAScript-style canonical numbers
  (`1.0` -> `1`, `0.7` stays `0.7`, exponent forms normalized to `e-7` /
  `e+21`). Full ECMAScript exponent formatting for extreme magnitudes is
  pinned at the 1.0.0 conformance freeze, exactly as in the Python binding.
  """

  @doc "The canonical RFC 8785 serialization of a JSON value, as a binary."
  def encode(nil), do: "null"
  def encode(true), do: "true"
  def encode(false), do: "false"
  def encode(n) when is_number(n), do: encode_number(n)
  def encode(s) when is_binary(s), do: encode_string(s)

  def encode(list) when is_list(list),
    do: "[" <> Enum.map_join(list, ",", &encode/1) <> "]"

  def encode(map) when is_map(map) do
    pairs =
      map
      |> Map.to_list()
      # Sort keys by code points, mirroring the Python reference's
      # sorted(..., key=[ord(c) for c in key]) — identical for ASCII keys.
      |> Enum.sort_by(fn {k, _v} -> String.to_charlist(k) end)

    "{" <>
      Enum.map_join(pairs, ",", fn {k, v} -> encode_string(k) <> ":" <> encode(v) end) <>
      "}"
  end

  def encode(other), do: raise(ArgumentError, "cannot canonicalize #{inspect(other)}")

  # ----------------------------------------------------------------- strings

  @doc "A JSON string with RFC 8785 minimal escaping."
  def encode_string(s) when is_binary(s) do
    inner = s |> String.to_charlist() |> Enum.map(&escape_char/1)
    IO.iodata_to_binary([?", inner, ?"])
  end

  defp escape_char(?"), do: "\\\""
  defp escape_char(?\\), do: "\\\\"
  defp escape_char(8), do: "\\b"
  defp escape_char(9), do: "\\t"
  defp escape_char(10), do: "\\n"
  defp escape_char(12), do: "\\f"
  defp escape_char(13), do: "\\r"

  defp escape_char(c) when c < 0x20 do
    # Lowercase four-digit hex, mirroring Python's "\\u%04x".
    hex = c |> Integer.to_string(16) |> String.downcase() |> String.pad_leading(4, "0")
    "\\u" <> hex
  end

  defp escape_char(c), do: <<c::utf8>>

  # ----------------------------------------------------------------- numbers

  @doc "ECMAScript-style canonical number serialization (RFC 8785)."
  def encode_number(n) when is_integer(n), do: Integer.to_string(n)

  def encode_number(n) when is_float(n) do
    # NaN and Infinity cannot arise: Elixir floats are always finite and our
    # own JSON parser only produces finite literals, so no isfinite gate is
    # needed here (the Python reference raises on them defensively).
    cond do
      n == 0.0 ->
        "0"

      trunc(n) == n and abs(n) < 1.0e21 ->
        # An integral float below 1e21 serializes as an integer (1.0 -> "1").
        Integer.to_string(trunc(n))

      true ->
        shortest(n)
    end
  end

  # The shortest round-trip decimal (Ryu via OTP's :short option), with the
  # exponent form normalized to ECMAScript style: mantissa without a
  # trailing ".0", exponent as e-7 (not e-07) or e+21.
  defp shortest(n) do
    text = :erlang.float_to_binary(n, [:short])

    case String.split(text, "e") do
      [_plain] -> text
      [mantissa, exponent] -> normalize_exponent(mantissa, exponent)
    end
  end

  defp normalize_exponent(mantissa, exponent) do
    mantissa =
      if String.ends_with?(mantissa, ".0") do
        binary_part(mantissa, 0, byte_size(mantissa) - 2)
      else
        mantissa
      end

    {sign, digits} =
      case exponent do
        "-" <> d -> {"-", d}
        "+" <> d -> {"+", d}
        d -> {"+", d}
      end

    digits =
      case String.trim_leading(digits, "0") do
        "" -> "0"
        d -> d
      end

    mantissa <> "e" <> sign <> digits
  end
end

File: bindings/ruby/lib/causalontology/jcs.rb
# frozen_string_literal: false

# RFC 8785 (JSON Canonicalization Scheme) serialization.
#
# Sorted keys (code-point order), minimal string escaping, ECMAScript-style
# canonical numbers (1.0 -> "1", 0.7 stays "0.7", exponent "1e-7" not
# "1e-07"). The number serialization implements the RFC 8785 rules for the
# value ranges Causalontology uses (integers, integer-valued floats, and
# short decimals); full ECMAScript exponent formatting for extreme
# magnitudes is pinned at the 1.0.0 conformance freeze.
#
# Ruby specifics relied on here: JSON.parse keeps the integer-versus-decimal
# source distinction ("1" -> Integer, "1.0" -> Float), and Float#to_s prints
# the shortest round-trip decimal with the same decimal/exponent thresholds
# as Python's repr (1e16 and 1e-4), so only the exponent spelling needs
# normalizing ("1.0e-07" -> "1e-7").

module Causalontology
  module Jcs
    # The two-character escapes of RFC 8785 section 3.2.2.2.
    ESCAPES = {
      '"'  => '\"',
      "\\" => "\\\\",
      "\b" => "\\b",
      "\t" => "\\t",
      "\n" => "\\n",
      "\f" => "\\f",
      "\r" => "\\r",
    }.freeze

    module_function

    # A JSON string literal with minimal escaping: the seven two-character
    # escapes, \u00xx for remaining control characters, everything else
    # (including multibyte text) passed through as UTF-8.
    def string(s)
      parts = +'"'
      s.each_char do |ch|
        if ESCAPES.key?(ch)
          parts << ESCAPES[ch]
        elsif ch.ord < 0x20
          parts << format("\\u%04x", ch.ord)
        else
          parts << ch
        end
      end
      parts << '"'
      parts
    end

    # A canonical JSON number: integers verbatim; integer-valued floats
    # below 1e21 printed as integers; other floats in shortest round-trip
    # form with the exponent normalized to ES6 style ("1e-7", "1e+21").
    def number(n)
      return n.to_s if n.is_a?(Integer)
      raise ArgumentError, "NaN and Infinity are not permitted (RFC 8785)" unless n.finite?
      return "0" if n == 0
      return n.to_i.to_s if n == n.truncate && n.abs < 1e21
      r = n.to_s # shortest round-trip decimal
      if r.include?("e")
        mant, exp = r.split("e")
        mant = mant.sub(/\.0\z/, "") # Ruby prints "1.0e-07"; ES6 prints "1e-7"
        sign = exp.start_with?("-") ? "-" : "+"
        digits = exp.sub(/\A[+-]/, "").sub(/\A0+/, "")
        digits = "0" if digits.empty?
        r = mant + "e" + sign + digits
      end
      r
    end

    # The canonical serialization of a parsed JSON value (nil, true/false,
    # Integer, Float, String, Array, or Hash with String keys).
    def encode(value)
      case value
      when nil
        "null"
      when true
        "true"
      when false
        "false"
      when Integer, Float
        number(value)
      when String
        string(value)
      when Array
        "[" + value.map { |v| encode(v) }.join(",") + "]"
      when Hash
        items = value.keys.sort_by(&:codepoints)
        "{" + items.map { |k| string(k) + ":" + encode(value[k]) }.join(",") + "}"
      else
        raise TypeError, "cannot canonicalize #{value.class}"
      end
    end
  end
end
